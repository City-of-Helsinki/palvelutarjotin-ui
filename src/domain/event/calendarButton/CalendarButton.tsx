import * as Sentry from '@sentry/browser';
import { saveAs } from 'file-saver';
import { Button, IconCalendar } from 'hds-react';
import { createEvent, EventAttributes } from 'ics';
import { useTranslation } from 'next-i18next';
import React from 'react';

import {
  OccurrenceFieldsFragment,
  EventFieldsFragment,
  usePlaceQuery,
} from '../../../generated/graphql';
import useLocale from '../../../hooks/useLocale';
import getDateArray from '../../../utils/getDateArray';
import getDomain from '../../../utils/getDomain';
import getLocalisedString from '../../../utils/getLocalisedString';
import { ROUTES } from '../../app/routes/constants';
import { getPlaceFields } from '../../place/utils';

interface CalendarButtonProps {
  event: EventFieldsFragment;
  occurrence: OccurrenceFieldsFragment;
}

const CalendarButton: React.FC<CalendarButtonProps> = ({
  event,
  occurrence,
}) => {
  const { t } = useTranslation();
  const locale = useLocale();
  const placeId = occurrence.placeId || event.location?.id || '';

  const { data, loading } = usePlaceQuery({
    variables: { id: placeId },
  });

  const { streetAddress, addressLocality, name: locationName } = getPlaceFields(
    data?.place,
    locale
  );

  const downloadIcsFile = () => {
    if (event.id && occurrence.startTime) {
      const domain = getDomain();
      const icsEvent: EventAttributes = {
        description: t('event:info.textCalendarLinkDescription', {
          description: getLocalisedString(event.shortDescription || {}, locale),
          link: `${domain}/${locale}${ROUTES.EVENT_DETAILS.replace(
            ':id',
            event.id
          )}`,
        }),
        end: occurrence.endTime
          ? getDateArray(occurrence.endTime)
          : getDateArray(occurrence.startTime),
        location: [locationName, streetAddress, addressLocality]
          .filter((e) => e)
          .join(', '),
        productId: domain,
        start: getDateArray(occurrence.startTime),
        startOutputType: 'local',
        title: getLocalisedString(event.name || {}, locale),
      };
      createEvent(icsEvent, (error: Error | undefined, value: string) => {
        if (error) {
          Sentry.captureException(error);
        } else {
          const blob = new Blob([value], { type: 'text/calendar' });
          saveAs(blob, `event_${event.id?.replace(/:/g, '')}.ics`);
        }
      });
    }
  };

  return !loading ? (
    <Button
      onClick={downloadIcsFile}
      iconLeft={<IconCalendar />}
      variant="supplementary"
    >
      {t('event:occurrenceList.downloadToCalendar')}
    </Button>
  ) : null;
};

export default CalendarButton;
