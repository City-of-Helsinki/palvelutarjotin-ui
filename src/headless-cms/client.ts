/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { ApolloClient, InMemoryCache } from '@apollo/client';

const apolloClient = new ApolloClient({
  uri: process.env.NEXT_PUBLIC_CMS_BASE_URL,
  cache: new InMemoryCache(),
});
export default apolloClient;
