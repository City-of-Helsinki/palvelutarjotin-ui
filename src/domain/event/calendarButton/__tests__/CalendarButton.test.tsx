import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library//user-event';
import { render, screen, waitFor } from '@testing-library/react';
import * as ICS from 'ics';
import React from 'react';

import { OccurrenceFieldsFragment } from '../../../../generated/graphql';
import { createPlaceQueryMock } from '../../../../tests/apollo-mocks/placeMocks';
import {
  fakeEvent,
  fakeLocalizedObject,
  fakePlace,
} from '../../../../utils/mockDataUtils';
import CalendarButton from '../CalendarButton';

const placeId = 'placeid-234324';

const eventMock = fakeEvent({
  location: fakePlace({ id: placeId }),
});

const mocks = [
  createPlaceQueryMock({
    id: placeId,
    name: fakeLocalizedObject('Lasten liikennekaupunki'),
  }),
];

const originalCreateEvent = ICS.createEvent;

beforeAll(() => {
  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
  (ICS as any).createEvent = jest.fn();
});

beforeEach(() => {
  jest.clearAllMocks();
});

afterAll(() => {
  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
  (ICS as any).createEvent = originalCreateEvent;
});

const occurrence: OccurrenceFieldsFragment = {
  startTime: '2020-07-15T09:00:00+00:00',
} as OccurrenceFieldsFragment;

test('matches snapshot', async () => {
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <CalendarButton occurrence={occurrence} event={eventMock} />
    </MockedProvider>
  );

  await screen.findByRole('button', {
    name: /Lataa kalenteriin/i,
  });

  expect(container.firstChild).toMatchSnapshot();
});

test('renders correct label', async () => {
  render(
    <MockedProvider mocks={mocks}>
      <CalendarButton occurrence={occurrence} event={eventMock} />
    </MockedProvider>
  );

  const downloadButton = await screen.findByRole('button', {
    name: /Lataa kalenteriin/i,
  });

  expect(downloadButton).toBeInTheDocument();
});

// FIXME: Fix the ICS-library usage in the tests
test.skip('works correctly when download button is clicked', async () => {
  render(
    <MockedProvider mocks={mocks}>
      <CalendarButton occurrence={occurrence} event={eventMock} />
    </MockedProvider>
  );

  const downloadButton = await screen.findByRole('button', {
    name: /Lataa kalenteriin/i,
  });

  await userEvent.click(downloadButton);

  expect(downloadButton).toBeInTheDocument();

  await waitFor(() => {
    expect(ICS.createEvent).toHaveBeenCalled();
  });
});
