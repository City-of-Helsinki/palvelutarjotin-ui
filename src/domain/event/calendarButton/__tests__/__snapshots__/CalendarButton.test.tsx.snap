// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<button
  class="Button-module_button__1msFE button_hds-button__2A0je Button-module_supplementary__3YKiS button_hds-button--supplementary__GcHcV"
  type="button"
>
  <div
    aria-hidden="true"
    class="Button-module_icon__O-h7R button_hds-icon__17j8Z"
  >
    <svg
      class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
      role="img"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <path
          d="M0 0h24v24H0z"
        />
        <path
          d="M17 2a1 1 0 011 1v1h4v17H2V4h4V3a1 1 0 112 0v1h8V3a1 1 0 011-1zm3 9H4v8h16v-8zm0-5H4v3h16V6z"
          fill="currentColor"
        />
      </g>
    </svg>
  </div>
  <span
    class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
  >
    Lataa kalenteriin...
  </span>
</button>
`;
